api/app.py

```diff
@@ -10,67 +13,257 @@
-            recipe['instructions'] = instructions if(instructions:= request.json.get('instructions')) else recipe['instructions']
-            recipe['categories'] = categories if(categories:= request.json.get('categories')) else recipe['categories']
-            recipe['ingredients'] = ingredients if(ingredients:= request.json.get('ingredients')) else recipe['ingredients']
-    return recipes[recipe_id - 1]
+    recipe = Recipe.query.get(recipe_id)
+
+    if not recipe:
+        return jsonify({'message': 'Recipe not found'}), 404
+
+    data = request.get_json()
+
+    if data is None:
+        return jsonify({'message': 'No input data provided'}), 400
+
+    categories = []
+    for category in data.get('categories', []):
+        name = category.get('name')
+        color = category.get('color', '#808080')
+        cat = Category.query.filter_by(name=name).first()
+        if not cat:
+            cat = Category(name=name, color=color)
+            db.session.add(cat)
+        categories.append(cat)
+
+    for ing in recipe.ingredients:
+        db.session.delete(ing)
+    db.session.flush()
+
+    ingredients = []
+    ingredient_regex = re.compile(r'^(?P<quantity>\d+)(?P<unit>[a-zA-Z]*) (?P<name>.+)$')
+    for ingredient in data.get('ingredients', []):
+        match = ingredient_regex.match(ingredient.strip())
+        if not match:
+            return jsonify({'error': f'Invalid ingredient format: {ingredient}'}), 400
+        ingredient = Ingredient(
+            name=match['name'],
+            unit=match['unit'] or None,
+            quantity=float(match['quantity']),
+            recipe=recipe
+        )
+        ingredients.append(ingredient)
+
+    recipe.name = data.get('name')
+    recipe.duration = data.get('duration')
+    recipe.pictures = data.get('pictures')
+    recipe.instructions = data.get('instructions')
+    recipe.categories = categories
+    recipe.ingredients = ingredients
+
+    db.session.commit()
+
+    return {
+        'id': recipe.id,
+        'name': recipe.name,
+        'duration': recipe.duration,
+        'pictures': recipe.pictures,
+        'instructions': recipe.instructions,
+        'categories': [category.name for category in recipe.categories],
+        'ingredients': [ingredient.name for ingredient in recipe.ingredients],
+    }
+
 
 @app.route('/api/recipes/<int:recipe_id>', methods=['DELETE'])
 def delete_recipe(recipe_id):
-    for recipe in recipes:
-        if(recipe['id'] == recipe_id):
-            recipes.remove(recipe)
-            return 'deleted recipe', 200
-    return 'Recipe not found', 404
+    recipe = Recipe.query.get(recipe_id)
+
+    if not recipe:
+        return jsonify({'error': 'Recipe not found'}), 404
+
+    for ingredient in recipe.ingredients:
+        db.session.delete(ingredient)
+
+    db.session.delete(recipe)
+    db.session.commit()
+
+    return jsonify({'message': 'Recipe deleted'}), 200
+
+def category_json(category):
```

Comment:
By adding the same `as_dict` method you used in `recipe.py` to `category.py`, we remove the necessity of this function.